### ISSUE TITLE
[bug] Incorrect handling of transitive dependencies when using version ranges instead of fixed versions

### ISSUE THREAD

**Issue by Ariox41** [conan-io/conan#14718]
### Environment details

* Operating System+version: ubuntu 22.04
* Compiler+version: gcc 11
* Conan version: 2.0.10
* Python version:  3.10.12

### Steps to reproduce

This issue is a continuation of https://github.com/conan-io/conan/issues/14695

The test is hosted on github [https://github.com/Ariox41/conan-transitive-deps-test](https://github.com/Ariox41/conan-transitive-deps-test).

The test creates the following simple graph:

```python
    req_version = '0.1.0'

    #Creates a package with package_type="library"
    util = ctx.library("util", '0.1.0')

    lib_a = ctx.library("lib_a", '0.1.0') \
        .requires(util, req_version, transitive_headers=None)
          
    lib_c = ctx.library("lib_c", '0.1.0') \
        .test_requires(lib_a, req_version)  \
        .test_requires(util, req_version ) 
```

When `req_version` is set to `"0.1.0"` - everything works fine. But if you set `req_version="[^0.1.0]"`, the header `util.hpp` becomes unavailable for `lib_c`. When debugging, you can see that for the `util` dependency, the value `require.headers` is set to `False`.

If you swap the `lib_a` and `util` dependencies, the test will also complete successfully:
```python
lib_c = ctx.library("lib_c", '0.1.0') \
        .test_requires(util, req_version ) \
        .test_requires(lib_a, req_version)  
```

> Note: the test uses `python_requires_extend`, this does not affect the result

> Upd: The test generates directories `build/utils`, `build/lib_a` and `build/lib_c` , the contents of these directories can be more clearly.

**Comment by AbrilRBS** [comment]
Hi @Ariox41 thanks a lot for your report.
I have thus far been unable to reproduce your issue outside the provided script. My current setup looks like this:

```py
def test_require_ranges():
  tc = TestClient()
  
  with tc.chdir("utils"):
      tc.run("new cmake_lib -d name=utils -d version=0.1.0")
      tc.run("create .")
  
  with tc.chdir("liba"):
      tc.run("new cmake_lib -d name=liba -d version=0.1.0 -d requires=utils/[>=0.1.0]")
      tc.run("create .")
  
  with tc.chdir("libc"):
      tc.run("new cmake_lib -d name=libc -d version=0.1.0 -d requires=liba/[>=0.1.0] -d requires=utils/[>=0.1.0]")
      new_conanfile = re.sub("self.requires",
                             "self.test_requires",
                             load(os.path.join(tc.current_folder, "conanfile.py")))
      tc.save({"conanfile.py": new_conanfile})
  
      # As per the report, this should fail when compiling libc.cpp, but it doesnt, it can find utils.h
      tc.run("create . -tf=", assert_error=True)
```

The last "conan create ." is executing succesfully, but as per your report, it should not.

Might this be something extra that is being done in your script's side? Maybe something to do with the pythonrequires?

I think that if you could reproduce the issue using "plain" Conan (ie, runing `conan news`, giving a more minimal plain exmaples), it would help a lot :)

**Comment by Ariox41** [comment]
I'm sorry, I didn't have time to simplify the example earlier. Here is a simplified example:
```python
import os
import subprocess
import shutil
from typing import List
import textwrap

build_folder = os.path.join(os.path.dirname(__file__), "build")
conan_cache_folder = os.path.join(build_folder, "conan2")


def main():
    os.environ["CONAN_HOME"] = conan_cache_folder
    shutil.rmtree(build_folder, ignore_errors=True)
    os.mkdir(build_folder)
    subprocess.run(["conan", "profile", "detect", "-f"],
                   cwd=build_folder).check_returncode()
    subprocess.run(["conan", "remote", "disable", "conancenter"],
                   cwd=build_folder).check_returncode()

    util_folder = os.path.join(build_folder, "util")
    lib_a_folder = os.path.join(build_folder, "lib_a")
    lib_b_folder = os.path.join(build_folder, "lib_b")
    os.mkdir(util_folder)
    os.mkdir(lib_a_folder)
    os.mkdir(lib_b_folder)

    with open(os.path.join(util_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile

class CMakeLibraryRecipe(ConanFile):
    name = "util"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
    generators = ["CMakeDeps"]
""")

    with open(os.path.join(lib_a_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile

class CMakeLibraryRecipe(ConanFile):
    name = "lib_a"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
    generators = ["CMakeDeps"]
    
    def requirements(self):
        self.requires("util/[>=0.1.0]")
""")
    with open(os.path.join(lib_b_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile

class CMakeLibraryRecipe(ConanFile):
    name = "lib_b"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
    generators = ["CMakeDeps"]
    
    def build_requirements(self):
        self.test_requires("lib_a/[>=0.1.0]")
        self.test_requires("util/[>=0.1.0]")
        
    def generate(self):
        util_items = []
        # The same dependencies are used by CMakeDeps
        for req, dep in list(self.dependencies.host.items()) + list(self.dependencies.build.items())+ list(self.dependencies.test.items()):
            if dep.ref.name == "util":
                util_items.append((req, dep))
        assert len(util_items) == 1, "Wrong test assumption" # ok
        req, dep = util_items[0]
        assert req.headers, "Incorrect behavior" # error
""")
    subprocess.run(["conan", "create", util_folder]).check_returncode()
    subprocess.run(["conan", "create", lib_a_folder]).check_returncode()
    subprocess.run(["conan", "create", lib_b_folder]).check_returncode()


if __name__ == "__main__":
    main()

```
The problem seems to appears  only for test_requires

**Comment by Ariox41** [comment]
Here is a more applied example of an error when using CMakeDeps:
```python
import os
import subprocess
import shutil

build_folder = os.path.join(os.path.dirname(__file__), "build")
conan_cache_folder = os.path.join(build_folder, "conan2")


def main():
    os.environ["CONAN_HOME"] = conan_cache_folder
    shutil.rmtree(build_folder, ignore_errors=True)
    os.mkdir(build_folder)
    subprocess.run(["conan", "profile", "detect", "-f"],
                   cwd=build_folder).check_returncode()
    subprocess.run(["conan", "remote", "disable", "conancenter"],
                   cwd=build_folder).check_returncode()

    util_folder = os.path.join(build_folder, "util")
    lib_a_folder = os.path.join(build_folder, "lib_a")
    lib_b_folder = os.path.join(build_folder, "lib_b")
    os.mkdir(util_folder)
    os.mkdir(lib_a_folder)
    os.mkdir(lib_b_folder)

    with open(os.path.join(util_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile

class CMakeLibraryRecipe(ConanFile):
    name = "util"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
""")

    with open(os.path.join(lib_a_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile

class CMakeLibraryRecipe(ConanFile):
    name = "lib_a"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
    
    def requirements(self):
        self.requires("util/[>=0.1.0]")
""")
    with open(os.path.join(lib_b_folder, "conanfile.py"), 'w') as f:
        f.write("""
from conan import ConanFile
from conan.tools.cmake import CMakeDeps, CMakeToolchain, cmake_layout

class CMakeLibraryRecipe(ConanFile):
    name = "lib_b"
    version = "0.1.0"
    package_type="library"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False], "fPIC": [True, False]}
    default_options = {"shared": True, "fPIC": True}
    
    def build_requirements(self):
        self.test_requires("lib_a/[>=0.1.0]")
        self.test_requires("util/[>=0.1.0]")
        
    def layout(self):
        cmake_layout(self)
    def generate(self):
        CMakeToolchain(self).generate()
        CMakeDeps(self).generate()
""")
    subprocess.run(["conan", "create", util_folder]).check_returncode()
    subprocess.run(["conan", "create", lib_a_folder]).check_returncode()
    subprocess.run(["conan", "build", lib_b_folder]).check_returncode()

    with open(os.path.join(lib_b_folder, "build", "Release", "generators", "util-release-x86_64-data.cmake")) as f:
        for line in f:
            if 'INCLUDE_DIRS_RELEASE' in line and ("set(" in line or "list(APPEND" in line):
                assert "conan2/p/" in line, f"util include path not found in '{line}'"

if __name__ == "__main__":
    main()
```

**Comment by AbrilRBS** [comment]
Thanks a lot for the report! 🫶

I was missing the libraries being shared in all instances for this to be reproducible. The new reproduction steps were super helpful and after some debugging with @memsharded we've found a solution and will release a bugfix in the next 2.0.11 to be released this week:tm:

**Comment by memsharded** [comment]
#14721 has been merged for next 2.0.11 that will solved this issue, thanks again for the feedback!
